p2p/peer.go

```diff
@@ -0,0 +1,218 @@
+/**
+*  @file
+*  @copyright defined in go-seele/LICENSE
+ */
+
+package p2p
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/seeleteam/go-seele/p2p/discovery"
+)
+
+type Peer struct {
+	fd       net.Conn
+	node     *discovery.Node
+	created  uint64
+	err      error
+	closed   chan struct{}
+	disc     chan uint
+	protoMap map[uint16]*Protocol // protoCode=>proto
+	capMap   map[string]uint16    //cap of protocol => protoCode
+
+	wMutex sync.Mutex // for conn write
+	wg     sync.WaitGroup
+}
+
+func (p *Peer) run() {
+	// add peer to protocols
+	for _, proto := range p.protoMap {
+		proto.AddPeerCh <- p
+	}
+	var (
+		writeErr = make(chan error, 1)
+		readErr  = make(chan error, 1)
+		//reason   uint // sent to the peer
+		err error
+	)
+	p.wg.Add(2)
+	go p.readLoop(readErr)
+	go p.pingLoop()
+
+	// Wait for an error or disconnect.
+loop:
+	for {
+		select {
+		case err = <-writeErr:
+			// A write finished. Allow the next write to start if
+			// there was no error.
+			if err != nil {
+				//reason = DiscNetworkError
+				break loop
+			}
+		case err = <-readErr:
+			p.err = err
+			break loop
+		//case err = <-p.disc:
+		case <-p.disc:
+			break loop
+		}
+	}
+
+	close(p.closed)
+	p.fd.Close()
+	p.wg.Wait()
+	// del peer from protocols
+	for _, proto := range p.protoMap {
+		proto.DelPeerCh <- p
+	}
+}
+
+func (p *Peer) pingLoop() {
+	ping := time.NewTimer(pingInterval)
+	defer p.wg.Done()
+	defer ping.Stop()
+	for {
+		select {
+		case <-ping.C:
+			p.sendCtlMsg(3)
+			ping.Reset(pingInterval)
+		case <-p.closed:
+			return
+		}
+	}
+}
+
+func (p *Peer) readLoop(errc chan<- error) {
+	defer p.wg.Done()
+	for {
+		msgRecv, err := p.recvRawMsg()
+		if err != nil {
+			errc <- err
+			return
+		}
+		if err = p.handle(msgRecv); err != nil {
+			errc <- err
+			return
+		}
+	}
+}
+
+func (p *Peer) handle(msgRecv *msg) error {
+	proto, ok := p.protoMap[msgRecv.protoCode]
+	if ok {
+		select {
+		case proto.ReadMsgCh <- &(msgRecv.Message):
+			return nil
+		case <-p.closed:
+			return io.EOF
+		}
+		return nil
+	}
+
+	if msgRecv.protoCode != 1 {
+		return errors.New("not valid protoCode")
+	}
+	// for control msg
+	switch {
+	case msgRecv.msgCode == ctlMsgPingCode:
+		go p.sendCtlMsg(ctlMsgPongCode)
+	case msgRecv.msgCode == ctlMsgDiscCode:
+		//var reason [1]DiscReason
+
+		//rlp.Decode(msg.Payload, &reason)
+		return fmt.Errorf("error=%d", ctlMsgDiscCode)
+	}
+	return nil
+}
+
+// SendMsg called by protocols
+func (p *Peer) SendMsg(proto *Protocol, msgSend *Message) error {
+	protoCode, ok := p.capMap[proto.cap().String()]
+	if !ok {
+		return errors.New("Not Found protoCode")
+	}
+	msgRaw := &msg{
+		protoCode: protoCode,
+		Message:   *msgSend,
+	}
+	return p.sendRawMsg(msgRaw)
+}
+
+func (p *Peer) sendCtlMsg(msgCode uint16) error {
```

Comment:
nice to rename this func name to sendControlMsg as Ctl is little meaningful for someone.